File: apps/fnos/backend/src/safe-save-contract.js
const CONTRACT_VERSION = 1;

const REJECTION_REASONS = Object.freeze([
  'conflict',
  'permission',
  'invalidTarget',
  'tooLarge',
  'invalidUTF8',
  'resourceExhausted',
  'unavailable',
  'cancelled',
  'internal',
]);

const CONFLICT_REASONS = new Set([
  'FILE_CONFLICT',
  'FILE_CHANGED_DURING_SAVE',
  'PATH_CHANGED_DURING_AUTHORIZATION',
  'PATH_CHANGED_DURING_OPEN',
  'PATH_CHANGED_DURING_SAVE',
  'OPENED_FD_RESOLUTION_FAILED',
  'SHARED_AUTHORIZATION_CHANGED',
  'RECOVERY_TARGET_CHANGED',
  'RECOVERY_IN_PROGRESS',
]);

const PERMISSION_REASONS = new Set([
  'USER_ACL_WRITE_DENIED',
  'USER_ACL_DENIED',
  'NO_AUTHORIZED_PATH',
  'PATH_NOT_AUTHORIZED',
  'PATH_OPEN_DENIED',
  'STORAGE_WRITE_DENIED',
]);

const INVALID_TARGET_REASONS = new Set([
  'INVALID_PATH',
  'INVALID_TARGET_TYPE',
  'INVALID_CONTENT',
  'INVALID_EXPECTED_REVISION',
  'INVALID_RECOVERY_ID',
  'INVALID_RECOVERY_VERSION',
  'PATH_NOT_FOUND',
  'RECOVERY_NOT_FOUND',
  'SYMLINK_SAVE_DENIED',
  'UNSUPPORTED_DOCUMENT_TYPE',
]);

const RESOURCE_EXHAUSTED_REASONS = new Set([
  'RECOVERY_BASELINE_TOO_LARGE',
  'RECOVERY_QUOTA_EXCEEDED',
  'STORAGE_FULL',
]);

const UNAVAILABLE_REASONS = new Set([
  'PATH_OPEN_FAILED',
  'PATH_OPEN_UNAVAILABLE',
  'PRECISE_FILE_STATE_UNAVAILABLE',
  'RECOVERY_STORAGE_UNAVAILABLE',
  'SECURE_FD_ACL_UNAVAILABLE',
  'SECURE_FD_PATH_UNAVAILABLE',
  'SHARED_AUTHORIZATION_FAILED',
  'STORAGE_WRITE_UNAVAILABLE',
]);

function validVersion(value) {
  return typeof value === 'string' && value.length > 0 ? value : null;
}

function validRecoveryId(value) {
  return typeof value === 'string' && /^[a-f0-9]{48}$/u.test(value)
    ? value
    : null;
}

function classifySaveError(error) {
  const reason = typeof error?.reason === 'string' ? error.reason : '';
  if (error?.name === 'AbortError') return 'cancelled';
  if (reason === 'FILE_TOO_LARGE') return 'tooLarge';
  if (reason === 'INVALID_UTF8') return 'invalidUTF8';
  if (CONFLICT_REASONS.has(reason)) return 'conflict';
  if (PERMISSION_REASONS.has(reason)) return 'permission';
  if (INVALID_TARGET_REASONS.has(reason)) return 'invalidTarget';
  if (RESOURCE_EXHAUSTED_REASONS.has(reason)) return 'resourceExhausted';
  if (UNAVAILABLE_REASONS.has(reason)) return 'unavailable';

  if (['EACCES', 'EPERM', 'EROFS'].includes(error?.code)) return 'permission';
  if (['ENOSPC', 'EDQUOT'].includes(error?.code)) return 'resourceExhausted';
  if (['EMFILE', 'ENFILE', 'ENOMEM', 'ESTALE', 'EBUSY', 'ETIMEDOUT'].includes(error?.code)) {
    return 'unavailable';
  }
  return 'internal';
}

function recoveryReferences(error, publicRecovery) {
  const recovery = publicRecovery ?? error?.recovery;
  const recoveryId = validRecoveryId(recovery?.recoveryId);
  if (!recoveryId) return [];
  return [{
    kind: 'privateJournal',
    reference: recoveryId,
    ...(typeof recovery.phase === 'string' && recovery.phase
      ? { phase: recovery.phase }
      : {}),
  }];
}

function committedSaveOutcome(result, { locator, includeContent = false } = {}) {
  const content = typeof result?.content === 'string' ? result.content : null;
  const byteCount = Number.isSafeInteger(result?.size) && result.size >= 0
    ? result.size
    : Buffer.byteLength(content ?? '', 'utf8');
  // A committed save must never be turned into an HTTP failure merely because
  // an adapter cannot read optional metadata. Production writes always return
  // revision; this fallback keeps the already-committed result honest.
  const version = validVersion(result?.revision)
    ?? `committed:${String(result?.mtime ?? 'unknown')}:${byteCount}`;
  const recovery = result?.recoveryCleanupPending
    ? [{ kind: 'cleanupPending', reference: 'fnos-private-journal-cleanup' }]
    : [];
  const snapshot = {
    locator: typeof locator === 'string' && locator ? locator : 'authorized-document',
    version,
    contentIncluded: Boolean(includeContent && content !== null),
    byteCount,
    capabilities: {
      readable: true,
      writable: result?.writable !== false,
      supportsCreate: false,
      supportsSaveAs: false,
    },
    implementationSemantics: {
      writeVisibility: 'recoverableInPlace',
      recoveryLocation: 'private',
    },
  };
  if (includeContent && content !== null) snapshot.content = content;
  return {
    contractVersion: CONTRACT_VERSION,
    kind: 'committed',
    snapshot,
    recoveryReferences: recovery,
  };
}

function failedSaveOutcome(error, { publicRecovery } = {}) {
  const currentVersion = validVersion(error?.currentRevision);
  if (error?.recoveryRequired || error?.reason === 'SAVE_RECOVERY_REQUIRED') {
    const phase = publicRecovery?.phase ?? error?.recovery?.phase;
    const commitState = phase === 'committed'
      ? 'committed'
      : phase === 'prepared'
        ? 'notCommitted'
        : 'unknown';
    return {
      contractVersion: CONTRACT_VERSION,
      kind: 'recoveryRequired',
      commitState,
      recoveryReferences: recoveryReferences(error, publicRecovery),
      ...(currentVersion ? { currentVersion } : {}),
    };
  }
  return {
    contractVersion: CONTRACT_VERSION,
    kind: 'rejected',
    reason: classifySaveError(error),
    ...(currentVersion ? { currentVersion } : {}),
  };
}

module.exports = {
  CONTRACT_VERSION,
  REJECTION_REASONS,
  classifySaveError,
  committedSaveOutcome,
  failedSaveOutcome,
};

File: apps/macos/FluxReader/Services/SafeSaveContract.swift
import CryptoKit
import Darwin
import Foundation

enum SafeSaveOutcomeKind: String, Codable, Equatable, Sendable {
  case committed
  case rejected
  case recoveryRequired
}

enum SafeSaveRejectionReason: String, Codable, CaseIterable, Equatable, Sendable {
  case conflict
  case permission
  case invalidTarget
  case tooLarge
  case invalidUTF8
  case resourceExhausted
  case unavailable
  case cancelled
  case `internal`
}

enum SafeSaveCommitState: String, Codable, Equatable, Sendable {
  case notCommitted
  case committed
  case unknown
}

enum SafeSaveWriteVisibility: String, Codable, Equatable, Sendable {
  case atomicReplace
  case recoverableInPlace
}

enum SafeSaveRecoveryLocation: String, Codable, Equatable, Sendable {
  case `private`
  case sidecar
}

struct SafeSaveCapabilities: Codable, Equatable, Sendable {
  let readable: Bool
  let writable: Bool
  let supportsCreate: Bool
  let supportsSaveAs: Bool
}

struct SafeSaveImplementationSemantics: Codable, Equatable, Sendable {
  let writeVisibility: SafeSaveWriteVisibility
  let recoveryLocation: SafeSaveRecoveryLocation
}

struct SafeSaveSnapshot: Codable, Equatable, Sendable {
  let locator: String
  let version: String
  let content: String?
  let contentIncluded: Bool
  let byteCount: Int
  let capabilities: SafeSaveCapabilities
  let implementationSemantics: SafeSaveImplementationSemantics
}

struct SafeSaveRecoveryReference: Codable, Equatable, Sendable {
  enum Kind: String, Codable, Equatable, Sendable {
    case privateJournal
    case retainedSidecar
    case cleanupPending
  }

  let kind: Kind
  let reference: String
  let phase: String?
}

struct SafeSaveOutcome: Codable, Equatable, Sendable {
  static let contractVersion = 1

  let contractVersion: Int
  let kind: SafeSaveOutcomeKind
  let snapshot: SafeSaveSnapshot?
  let reason: SafeSaveRejectionReason?
  let commitState: SafeSaveCommitState?
  let recoveryReferences: [SafeSaveRecoveryReference]
  let currentVersion: String?
}

enum SafeSaveContract {
  static func committed(
    document: MarkdownDocument,
    locator: String? = nil,
    includeContent: Bool = false
  ) -> SafeSaveOutcome {
    SafeSaveOutcome(
      contractVersion: SafeSaveOutcome.contractVersion,
      kind: .committed,
      snapshot: SafeSaveSnapshot(
        locator: locator ?? document.url.standardizedFileURL.absoluteString,
        version: versionToken(for: document),
        content: includeContent ? document.content : nil,
        contentIncluded: includeContent,
        byteCount: document.byteCount,
        capabilities: SafeSaveCapabilities(
          readable: true,
          writable: true,
          supportsCreate: true,
          supportsSaveAs: true
        ),
        implementationSemantics: SafeSaveImplementationSemantics(
          writeVisibility: .atomicReplace,
          recoveryLocation: .sidecar
        )
      ),
      reason: nil,
      commitState: nil,
      recoveryReferences: [],
      currentVersion: nil
    )
  }

  static func failed(error: Error) -> SafeSaveOutcome {
    if case .saveRecoveryRequired(let path) = error as? FileAccessError {
      return SafeSaveOutcome(
        contractVersion: SafeSaveOutcome.contractVersion,
        kind: .recoveryRequired,
        snapshot: nil,
        reason: nil,
        commitState: .unknown,
        recoveryReferences: [
          SafeSaveRecoveryReference(
            kind: .retainedSidecar,
            reference: opaqueReference(for: path),
            phase: "recovery-required"
          )
        ],
        currentVersion: nil
      )
    }

    return SafeSaveOutcome(
      contractVersion: SafeSaveOutcome.contractVersion,
      kind: .rejected,
      snapshot: nil,
      reason: rejectionReason(for: error),
      commitState: nil,
      recoveryReferences: [],
      currentVersion: nil
    )
  }

  static func versionToken(for document: MarkdownDocument) -> String {
    var material = Data(document.content.utf8)
    material.append(0)
    material.append(
      Data(
        "\(document.byteCount):\(document.modificationDate?.timeIntervalSince1970.bitPattern ?? 0)"
          .utf8
      )
    )
    return "macos:\(hexDigest(material))"
  }

  private static func rejectionReason(for error: Error) -> SafeSaveRejectionReason {
    if error is CancellationError { return .cancelled }

    if let accessError = error as? FileAccessError {
      switch accessError {
      case .unsupportedFileType, .notRegularFile:
        return .invalidTarget
      case .fileTooLarge:
        return .tooLarge
      case .invalidUTF8:
        return .invalidUTF8
      case .fileModifiedExternally:
        return .conflict
      case .saveRecoveryRequired:
        // Handled before this function; retain a fail-closed classification if
        // a future call site bypasses that branch.
        return .internal
      }
    }

    let nsError = error as NSError
    if nsError.domain == NSPOSIXErrorDomain {
      switch Int32(nsError.code) {
      case EACCES, EPERM, EROFS:
        return .permission
      case ENOSPC, EDQUOT:
        return .resourceExhausted
      case EMFILE, ENFILE, ENOMEM, ESTALE, EBUSY, ETIMEDOUT:
        return .unavailable
      default:
        break
      }
    }

    if nsError.domain == NSCocoaErrorDomain {
      switch nsError.code {
      case CocoaError.Code.fileReadNoPermission.rawValue,
        CocoaError.Code.fileWriteNoPermission.rawValue,
        CocoaError.Code.fileWriteVolumeReadOnly.rawValue:
        return .permission
      case CocoaError.Code.fileWriteOutOfSpace.rawValue:
        return .resourceExhausted
      case CocoaError.Code.fileWriteFileExists.rawValue:
        return .conflict
      case CocoaError.Code.fileNoSuchFile.rawValue:
        return .invalidTarget
      default:
        break
      }
    }

    if let underlying = nsError.userInfo[NSUnderlyingErrorKey] as? Error {
      return rejectionReason(for: underlying)
    }
    return .internal
  }

  private static func opaqueReference(for value: String) -> String {
    "sidecar:\(hexDigest(Data(value.utf8)))"
  }

  private static func hexDigest(_ data: Data) -> String {
    SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
  }
}
